### ISSUE TITLE
how to deploy a dolt cluster in k8s?

### ISSUE THREAD

**Issue by chicco785** [dolthub/doltclusterctl#1]
hi!
this project looks interesting, and as far as I understood help to manage ha with active / passive deployment of a dolt cluster. I couldn't find though any information on how to actually deploy the dolt cluster :) any hint will be welcome!

**Comment by reltuk** [comment]
Hi @chicco785! Thanks for your interest in dolt.

We don't currently have published guidance on running a cluster in kubernetes. I can provide some guidance on how we've typically gone about it recently, but I can't guarantee it will meet all your use cases.

We typically deploy it as a statically sized `StatefulSet`, with a unique rendered config file for each pod instance. Those config files each configure the that pod to replicate to the other pods in the StatefulSet.

We use three `Service`s to address and route traffic to the pods:

1) Is a [headless Service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) which allows the pods themselves to address each other directly using hostnames of the form `POD_NAME.INTERNAL_SERVICE_NAME.NAMESPACE...`. It is used in the cluster replication URLs which dolt is configured with.

2) Is a ClusterIP or LoadBalancer Service which targets the pod in the StatefulSet which is labeled as `primary`.

3) Is a ClusterIP or LoadBalancer Service which targets the pod(s) in the StatefulSet which are labeled 'standby'.

Currently `dolt` itself does not integrate with the kubernetes control plane, but traffic routing in the above setup relies on unique labels being maintained on the Pods within the cluster. So an external agent needs to run to apply the correct labels, and it needs to run to update the labels anytime there is a failover or a pod is deleted and recreated by the StatefulSet controller, for example.

I've scratched together some example manifests for possible deployment in a gist here: https://gist.github.com/reltuk/8a97771e8b46dd32e47c80ef0c3645f7

You can also see how some kubernetes API objects come together for some test deployments in the e2e tests of this project. For example, here is where we create a typical simple StatefulSet of pods which will replicate amongst each other: https://github.com/dolthub/doltclusterctl/blob/09d3026b3959c08ad33f17e4dac4a21c2d47c9de/e2e/statefulset_test.go#L185

Very happy to answer any further questions you have here.

As cluster replication becomes more mature, we hope to add more guidance to how to use it ergonomically from within a Kubernetes to our official documentation.

**Comment by chicco785** [comment]
> Hi @chicco785! Thanks for your interest in dolt.
> 
> We don't currently have published guidance on running a cluster in kubernetes. I can provide some guidance on how we've typically gone about it recently, but I can't guarantee it will meet all your use cases.
> 
> We typically deploy it as a statically sized `StatefulSet`, with a unique rendered config file for each pod instance. Those config files each configure the that pod to replicate to the other pods in the StatefulSet.
> 
> We use three `Service`s to address and route traffic to the pods:
> 
> 1. Is a [headless Service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) which allows the pods themselves to address each other directly using hostnames of the form `POD_NAME.INTERNAL_SERVICE_NAME.NAMESPACE...`. It is used in the cluster replication URLs which dolt is configured with.
> 2. Is a ClusterIP or LoadBalancer Service which targets the pod in the StatefulSet which is labeled as `primary`.
> 3. Is a ClusterIP or LoadBalancer Service which targets the pod(s) in the StatefulSet which are labeled 'standby'.
> 
> Currently `dolt` itself does not integrate with the kubernetes control plane, but traffic routing in the above setup relies on unique labels being maintained on the Pods within the cluster. So an external agent needs to run to apply the correct labels, and it needs to run to update the labels anytime there is a failover or a pod is deleted and recreated by the StatefulSet controller, for example.
> 
> I've scratched together some example manifests for possible deployment in a gist here: https://gist.github.com/reltuk/8a97771e8b46dd32e47c80ef0c3645f7
> 
> You can also see how some kubernetes API objects come together for some test deployments in the e2e tests of this project. For example, here is where we create a typical simple StatefulSet of pods which will replicate amongst each other:
> 
> https://github.com/dolthub/doltclusterctl/blob/09d3026b3959c08ad33f17e4dac4a21c2d47c9de/e2e/statefulset_test.go#L185
> 
> Very happy to answer any further questions you have here.
> 
> As cluster replication becomes more mature, we hope to add more guidance to how to use it ergonomically from within a Kubernetes to our official documentation.

More than enough to start :) in case we are going to use dolt in our products we may have resources to deliver an helm chart to the community. Would that be of interest for you guys?

thx a lot!

**Comment by reltuk** [comment]
> More than enough to start :) in case we are going to use dolt in our products we may have resources to deliver an helm chart to the community. Would that be of interest for you guys?

Absolutely! We would love to collaborate on some Helm charts which could help people get started with Dolt in Kubernetes faster.

**Comment by priley86** [comment]
@reltuk - can you possibly expand on how you are using this library in a practical sense? It seems to have some powerful utilities for running dolt within Kubernetes w/ replication setups (active/active) or (active/passive).

Can you provide us with any advice or point to any documentation on utilizing this within an agent/sidecar that monitors things like failover? How are you triggering the utils here for things like applying labels, graceful failover, and rolling restarts? Is another sidecar handling this? If that's private IP that you can't share, certainly understand, but figured I'd ask here. 😄 

Context: i'm working on a simple two node setup like the one you've referenced here in a [fork](https://github.com/bitboxed/bitnode/tree/main), and i'd like to ensure I can support things like failover w/ Dolt seamlessly.

**Comment by reltuk** [comment]
@priley86 Sorry for the delay in a response here.

Unfortunately, we don't currently have a documented deployment or any publicly available examples of automated day-two operations, especially with regards to automated fail over from a primary to a standby. As I'm sure you've noticed, `doltclusterctl` itself is currently only positioned to perform the operations themselves, as a one-shot invocation.

At a high level, the way we currently use doltclusterctl is:

* In our initial deployments, we just one shot applyprimarylabels.

* When we are upgrading a StatefulSet, we use it to one shot the rolling restart, where the update policy on the statefulset itself is on-delete.

* If we have a particular need to drain a primary (node rotation, for example), we will use it to one shot the graceful failover.

However, we don't currently use `doltclusterctl` to promote standby in the event of a primary failure. For automated failover from a primary to a standby in the case of an unexpected failure, we actually have a different external agent that checks health of the current primary and coordinates to do the equivalent of a `promotestandby`. However, that agent is not currently open source as its somewhat coupled currently with our hosted.doltdb.com implementation.

I can also point out that the code in this repository behaves as expected when everything goes really well, but it can leave a bit to be desired in the case of partial failure, for example. So, rollingrestart, for example, could leave the cluster partially upgraded and require manual intervention depending upon how the restarts go.

The code in this repository has been adopted by some teams and developed into a full fledge k8s operator which does provisioning of replicated clusters and which also does active monitoring to determine when to attempt automated failovers. That operator is unfortunately not yet open source, but we hope to be able to open source it eventually. I will definitely keep you posted if and when that time arrives.

Until then, we're happy to be a point of reference if you have any questions about how to best proceed with preparing a Dolt deployment for high availability and making it production-ready. Sorry I don't have a better off the shelf solution right now. And thank you for your interest in Dolt!

**Comment by priley86** [comment]
@reltuk thanks - super helpful! 🙇 

yes, was reading about this operator in development in your proxysql blog post (also very insightful).

i'll be sure to report any feedback i can as i'm researching and will definitely stay tuned 👍
